scripts/binary/binary-cleanup.js

```diff
@@ -0,0 +1,190 @@
+    ...['ibmi',
+      'sunos',
+      'android',
+      'darwin',
+      'freebsd',
+      'linux',
+      'openbsd',
+      'sunos',
+      'win32'].map((platform) => require.resolve(`default-gateway/${platform}`)),
+  ])
+  let esbuildResult
+  let newEntryPointsFound = true
+
+  // The general idea here is to run esbuild with on entry points that are used outside of the snapshot. If, during the process,
+  // we find places where we do a require.resolve on a module, that should be treated as an additional entry point and we run
+  // esbuild again. We do this until we no longer find any new entry points. The resulting metafile inputs are
+  // the dependency paths that we need to ensure stay in the snapshot.
+  while (newEntryPointsFound) {
+    esbuildResult = await esbuild.build({
+      entryPoints: [...entryPoints],
+      bundle: true,
+      outdir: workingDir,
+      platform: 'node',
+      metafile: true,
+      external: [
+        './packages/server/server-entry',
+        'fsevents',
+        'pnpapi',
+        '@swc/core',
+        'emitter',
+      ],
+    })
+
+    newEntryPointsFound = false
+    esbuildResult.warnings.forEach((warning) => {
+      const matches = warning.text.match(/"(.*)" should be marked as external for use with "require.resolve"/)
+      const warningSubject = matches && matches[1]
+
+      if (warningSubject) {
+        let entryPoint
+
+        if (warningSubject.startsWith('.')) {
+          entryPoint = path.join(__dirname, '..', '..', path.dirname(warning.location.file), warningSubject)
+        } else {
+          entryPoint = require.resolve(warningSubject)
+        }
+
+        if (path.extname(entryPoint) !== '' && !entryPoints.has(entryPoint)) {
+          newEntryPointsFound = true
+          entryPoints.add(entryPoint)
+        }
+      }
+    })
+  }
+
+  return [...Object.keys(esbuildResult.metafile.inputs), ...entryPoints]
+}
+
+const cleanup = async (buildAppDir) => {
+  // 1. Retrieve all dependencies that still need to be kept in the binary. In theory, we could use the bundles generated here as single files within the binary,
+  // but for now, we just track on the dependencies that get pulled in
+  const keptDependencies = await getDependencyPathsToKeep()
+
+  // 2. Gather the dependencies that could potentially be removed from the binary due to being in the snapshot
+  const potentiallyRemovedDependencies = [...snapshotMetadata.healthy, ...snapshotMetadata.deferred, ...snapshotMetadata.norewrite]
+
+  // 3. Remove all dependencies that are in the snapshot but not in the list of kept dependencies from the binary
+  await Promise.all(potentiallyRemovedDependencies.map(async (dependency) => {
+    if (![...keptDependencies, 'package.json', 'packages/server/server-entry.js'].includes(dependency.slice(2)) && !dependency.includes('marionette-client')) {
```

Comment:
Maybe add the extra dependencies above when declaring `keptDependencies` instead of here.

Reply:
Fixed here: [`6680189` (#24262)](https://github.com/cypress-io/cypress/pull/24262/commits/66801890ef5989448eb17cbcef56b99f2ef328ab)